src/dev/pca9702.rs

```
@@ -0,0 +1,194 @@
+            in3: crate::Pin::new(3, &self.0),
+            in4: crate::Pin::new(4, &self.0),
+            in5: crate::Pin::new(5, &self.0),
+            in6: crate::Pin::new(6, &self.0),
+            in7: crate::Pin::new(7, &self.0),
+        }
+    }
+}
+
+/// Container for all 8 input pins on the PCA9702.
+pub struct Parts<'a, B, M = core::cell::RefCell<Driver<B>>>
+where
+    B: Pca9702BusTrait,
+    M: crate::PortMutex<Port = Driver<B>>,
+{
+    pub in0: crate::Pin<'a, crate::mode::Input, M>,
+    pub in1: crate::Pin<'a, crate::mode::Input, M>,
+    pub in2: crate::Pin<'a, crate::mode::Input, M>,
+    pub in3: crate::Pin<'a, crate::mode::Input, M>,
+    pub in4: crate::Pin<'a, crate::mode::Input, M>,
+    pub in5: crate::Pin<'a, crate::mode::Input, M>,
+    pub in6: crate::Pin<'a, crate::mode::Input, M>,
+    pub in7: crate::Pin<'a, crate::mode::Input, M>,
+}
+
+/// Internal driver struct for PCA9702.
+pub struct Driver<B> {
+    bus: B,
+}
+
+impl<B> Driver<B> {
+    fn new(bus: B) -> Self {
+        Self { bus }
+    }
+}
+
+/// Trait for the underlying PCA9702 SPI bus. Simpler than e.g. MCP23S17
+/// because PCA9702 is read-only and has no register-based protocol.
+pub trait Pca9702BusTrait {
+    type BusError;
+
+    /// Reads 8 bits from the device (which represent the state of inputs [in7..in0])
+    fn read_inputs(&mut self) -> Result<u8, Self::BusError>;
+}
+
+impl<B: Pca9702BusTrait> PortDriver for Driver<B> {
+    /// Our `Error` is a custom enum wrapping both bus errors and an unsupported-ops error.
+    type Error = Pca9702Error<B::BusError>;
+
+    /// PCA9702 is input-only, return an error here.
+    fn set(&mut self, _mask_high: u32, _mask_low: u32) -> Result<(), Self::Error> {
+        Err(Pca9702Error::OutputNotSupported)
+    }
+
+    /// PCA9702 is input-only, return an error here.
+    fn is_set(&mut self, _mask_high: u32, _mask_low: u32) -> Result<u32, Self::Error> {
+        Err(Pca9702Error::OutputNotSupported)
+    }
```

These should panic instead of returning an error.  Errors should be returned when "something went wrong"; panics should occur when the programmer made a mistake.

In this case, calling `set()` on an input-only expander can only occur by severe programmer error, so a panic is the right choice.